### ISSUE TITLE
Switch back from GIST to GIN

### ISSUE THREAD

**Issue by ara4n** [matrix-org/synapse#2753]
Currently we create FTS indexes as GIN in the schema but then convert to GIST as a background job in the main search storage (why isn't this a schema delta script?).  We want to try GIN again on matrix.org however, as with faster IO perhaps it will perform better for inserts.  If it doesn't, there are some tricks we can try short of upgrading postgres to 9.6 or later

**Comment by ara4n** [comment]
we just did this on matrix.org and it sped up search from 200,000ms to 6.71ms for a rare item (an eth account which appeared in 3 different accounts). so i think we should back out the bg conversion job...

**Comment by ara4n** [comment]
We're seeing some massive (>1 minute) pauses every N hours with GIN, as predicted above:

```
2018-01-08 16:44:32.874 GMT [matrix] LOG:  duration: 84165.911 ms  statement: INSERT INTO event_search (event_id, room_id, key, vector, stream_ordering, origin_server_ts) VALUES ('$15154297881052308GgHpJ:matrix.org','!yBVfMfMeyivGcJPbvS:matrix.org','content.body'
2018-01-08 16:44:32.874 GMT [matrix] LOG:  duration: 82618.283 ms  statement: INSERT INTO event_search (event_id, room_id, key, vector, stream_ordering, origin_server_ts) VALUES ('$15154297881052316hpNnS:matrix.org','!gMDwyoyuEhzHNjElAw:matrix.org','content.body'
2018-01-08 16:44:32.874 GMT [matrix] LOG:  duration: 23046.208 ms  statement: INSERT INTO event_search (event_id, room_id, key, vector, stream_ordering, origin_server_ts) VALUES ('$15154297891052324ZUwTS:matrix.org','!pBjZyyNmXwLCBPUAlE:matrix.org','content.body'
2018-01-08 16:44:32.874 GMT [matrix] LOG:  duration: 82105.092 ms  statement: INSERT INTO event_search (event_id, room_id, key, vector, stream_ordering, origin_server_ts) VALUES ('$15154297891052318FwGMB:matrix.org','!PVQRUJMgyQViHjyVgl:matrix.org','content.body'
```

Looks like we need to set work_mem to a small value at the point of doing the `INSER INTO event_search` as suggested.

**Comment by richvdh** [comment]
this got fixed by #2769, though as #4219 notes, we are still having problems.
